text/0000_unsafe_modules.md

```diff
@@ -0,0 +1,79 @@
+- Feature Name: unsafe_modules
+- Start Date: 2017-09-12
+- RFC PR: (leave this empty)
+- Rust Issue: (leave this empty)
+
+# Summary
+Allow unsafe at more positions.
+
+# Motivation
+
+When writing unsafe code it's annoying having to write unsafe everywhere. Normally that's a good thing, but sometimes, especially for prototyping, it would sometimes be preferable to be able to declare unsafe for larger blocks.
+Maybe it's also helpful for low level code with huge amount of unsafe functions or c interaction.
+
+# Guide-level explanation
+
+Unsafe blocks can also be written outside of functions:
+```
+unsafe {
+    fn x() {...}
+    trait X {...}
+    ...
+}
+```
+
+This will make the elements defined inside the block unsafe.
+It will be equivalent to following code:
+```
+unsafe fn x() {...}
+unsafe trait X {...}
+unsafe ...
+```
+
+Even modules can be declared as unsafe like this:
+```
+unsafe mod test {
+    fn x() {...}
+    trait X {...}
+    ...
+}
+```
+
+This is almost equivalent to following: 
+```
+mod test {
+    unsafe {
+        fn x() {...}
+        trait X {...}
+        ...
+    }
+}
+```
+But unsafe modules have an additional property.
+Using such a module is also unsafe:
+```
+unsafe use test;
```

Comment:
I find this surprising, because `use` doesn't need `unsafe` to import other `unsafe` things:
```rust
mod foo { pub unsafe fn bar() {} }
use foo::bar; // no unsafe
```

Reply:
Still perhaps understandable...? With normal modules the entire module isn't necessarily for unsafe stuff. And unsafe use can be good for greppability. Then again, you still have to be in "unsafe mode" to use the functions inside the module, and you can grow for that.

Reply:
It means that "this could just be implemented by some syntactic transformations" isn't true, though.

Reply:
I also wasn't sure if unsafe use is useful. But you don't even need to `use` unsafe functions from safe modules, and just use them at other positions. Maybe just importing things from inside the module as unsafe would make sence? But I think I should remove `unsafe` `use` from the RFC entirely. What do you think?

Reply:
I'd say just remove it, especially since it opens a bunch of questions about what paths involving the module would require.  For example:
```rust
unsafe mod foo {
    pub struct Bar;
}
use foo::Bar; // do I need `unsafe use` here?
```
(Thanks to Havvy for inspiration that led to this example.)